x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlIndexTemplateRegistry.java

```diff
@@ -0,0 +1,119 @@
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.MlConfigIndex;
+import org.elasticsearch.xpack.core.ml.MlMetaIndex;
+import org.elasticsearch.xpack.core.ml.inference.persistence.InferenceIndexConstants;
+import org.elasticsearch.xpack.core.ml.job.persistence.AnomalyDetectorsIndex;
+import org.elasticsearch.xpack.core.ml.job.persistence.AnomalyDetectorsIndexFields;
+import org.elasticsearch.xpack.core.ml.notifications.NotificationsIndex;
+import org.elasticsearch.xpack.core.template.IndexTemplateConfig;
+import org.elasticsearch.xpack.core.template.IndexTemplateRegistry;
+import org.elasticsearch.xpack.core.template.LifecyclePolicyConfig;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public class MlIndexTemplateRegistry extends IndexTemplateRegistry {
+
+    private static final String ROOT_RESOURCE_PATH = "/org/elasticsearch/xpack/core/ml/";
+    private static final String ANOMALY_DETECTION_PATH = ROOT_RESOURCE_PATH + "anomalydetection/";
+    private static final String VERSION_PATTERN = "xpack.ml.version";
+    private static final String VERSION_ID_PATTERN = "xpack.ml.version.id";
+
+    private static final IndexTemplateConfig ANOMALY_DETECTION_RESULTS_TEMPLATE = anomalyDetectionResultsTemplate();
+
+    private static final IndexTemplateConfig ANOMALY_DETECTION_STATE_TEMPLATE = new IndexTemplateConfig(
+        AnomalyDetectorsIndexFields.STATE_INDEX_PREFIX,ANOMALY_DETECTION_PATH + "state_index_template.json",
+        Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig META_TEMPLATE = new IndexTemplateConfig(MlMetaIndex.INDEX_NAME,
+        ROOT_RESOURCE_PATH + "meta_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig NOTIFICATIONS_TEMPLATE = new IndexTemplateConfig(NotificationsIndex.NOTIFICATIONS_INDEX,
+        ROOT_RESOURCE_PATH + "notifications_index_template.json", Version.CURRENT.toString(), VERSION_PATTERN,
+        Collections.singletonMap(Pattern.quote("${" + VERSION_ID_PATTERN + "}"), String.valueOf(Version.CURRENT.id)));
+
+    private static final IndexTemplateConfig CONFIG_TEMPLATE = configTemplate();
+
+    private static final IndexTemplateConfig INFERENCE_TEMPLATE = new IndexTemplateConfig(InferenceIndexConstants.LATEST_INDEX_NAME,
```

Comment:
This template's name is formed by the inference prefix plus the latest version. This would mean that once we change version we'll create a new template. I think that's probably not a good idea. I left it as is for now but we can address in this PR.